server/packet_handlers.py

```diff
@@ -229,59 +232,105 @@ async def ping_handler(session: "Session", packet_data: bytes):
-    
+
     packet_reader = packets.PacketReader(packet_data)
     host_account_id = packet_reader.read_i32()
 
     host_session = await sessions.fetch_by_account_id(host_account_id)
-    
+
     if host_session is None:
         logger.warning(
             "A user attempted to spectate another user who is offline",
             spectator_id=session["account_id"],
-            host_id=host_session["account_id"],
+            host_id=host_account_id,
         )
         return
 
-    host_account_UUID = await spectators.start_spectating(
-        host_account_id,
+    await spectators.add(
         host_session["session_id"],
+        session["session_id"],
     )
-    
+
+    # TODO: make this stateless
+    who_spectating_who[session["session_id"]] = host_session["session_id"]
+
     await packet_bundles.enqueue(
-        host_account_UUID,
-        packets.write_spectator_joined_packet(session["account_id"])
+        host_session["session_id"],
+        packets.write_spectator_joined_packet(session["account_id"]),
     )
 
+    for spectator_session_id in await spectators.members(host_session["session_id"]):
+        if spectator_session_id == session["session_id"]:
+            continue
 
+        await packet_bundles.enqueue(
+            spectator_session_id,
+            packets.write_fellow_spectator_joined_packet(session["account_id"]),
+        )
 
 
 # STOP_SPECTATING = 17
 
 
+@bancho_handler(packets.ClientPackets.STOP_SPECTATING)
+async def stop_spectating_handler(session: "Session", packet_data: bytes):
+    assert session["presence"] is not None
+
+    # TODO: make this stateless
+    host_session_id = who_spectating_who[session["session_id"]]
+    host_session = await sessions.fetch_by_id(host_session_id)
+
+    if host_session is None:
+        # logger.warning(
+        #     "A user attempted to stop spectating another user who is offline",
+        #     spectator_id=session["account_id"],
+        #     host_id=host_account_id,
+        # )
+        return
+
+    await spectators.remove(
+        host_session["session_id"],
+        session["session_id"],
+    )
+
+    # TODO: make this stateless
+    del who_spectating_who[session["session_id"]]
+
+    await packet_bundles.enqueue(
+        host_session["session_id"],
+        packets.write_spectator_left_packet(session["account_id"]),
+    )
+
+    for spectator_session_id in await spectators.members(host_session["session_id"]):
+        if spectator_session_id == session["session_id"]:
+            continue
+
+        await packet_bundles.enqueue(
+            spectator_session_id,
+            packets.write_fellow_spectator_left_packet(session["account_id"]),
+        )
+
+
 # SPECTATE_FRAMES = 18
 
+
 @bancho_handler(packets.ClientPackets.SPECTATE_FRAMES)
 async def spectate_frames_handler(session: "Session", packet_data: bytes):
-    packet_reader = packets.PacketReader(packet_data)
-    host_account_id = packet_reader.read_i32()
```

Comment:
this packet does not receive an account id

https://github.com/osuAkatsuki/bancho.py/blob/master/app/api/domains/cho.py#L1071